source/http/HttpConnectionManager.cpp

```diff
@@ -0,0 +1,349 @@
+                    if (poolOrVend && !m_pendingConnectionRequests.empty())
+                    {
+                        userCallback = m_pendingConnectionRequests.front();
+                        m_pendingConnectionRequests.pop_front();
+                        ++m_outstandingVendedConnections;
+                        callbackError = AWS_ERROR_SUCCESS;
+                        vendedConnection = std::move(connection);
+                    }
+                    else if (poolOrVend)
+                    {
+                        /* Case 2.2 */
+                        m_connections.push_back(std::move(connection));
+                    }
+                }
+
+                if (userCallback)
+                {
+                    userCallback(vendedConnection, callbackError);
+                }
+            }
+
+            void HttpClientConnectionManager::ReleaseConnection(
+                std::shared_ptr<HttpClientConnection> connection) noexcept
+            {
+                poolOrVendConnection(std::move(connection), true);
+            }
+
+            void HttpClientConnectionManager::onConnectionSetup(
+                const std::shared_ptr<HttpClientConnection> &connection,
+                int errorCode) noexcept
+            {
+                if (!errorCode && connection)
+                {
+                    poolOrVendConnection(connection, false);
+                    return;
+                }
+
+                /* this only gets hit if the connection setup failed. */
+                Vector<std::pair<OnClientConnectionAvailable, int>> userCallbackNewConnectionFailures;
+                {
+                    std::lock_guard<std::mutex> connectionsLock(m_connectionsLock);
+                    --m_pendingConnections;
+                    if (!m_pendingConnectionRequests.empty())
+                    {
+                        userCallbackNewConnectionFailures.push_back(std::pair<OnClientConnectionAvailable, int>(
+                            m_pendingConnectionRequests.front(), errorCode));
+                        m_pendingConnectionRequests.pop_front();
+                    }
+
+                    /* if we had more pending connection requests than the pool size, we need to replace this
+                       connection. Just loop until we've matched either the connectionRequest to the max amount of
+                       connections in flight, or when run out of pending connection requests. */
+                    while (m_pendingConnectionRequests.size() > m_pendingConnections &&
+                           m_connections.size() + m_pendingConnections + m_outstandingVendedConnections < m_maxSize)
+                    {
+                        if (!createConnection())
+                        {
+                            userCallbackNewConnectionFailures.push_back(std::pair<OnClientConnectionAvailable, int>(
+                                m_pendingConnectionRequests.front(), aws_last_error()));
```

Comment:
trivial: could use emplace_back(), and could std::move(m_pendingConnectionRequests.front())